### PR TITLE
feat: EXC-2062: Add System Api for environment variables

### PR DESCRIPTION
**Pull request by AlexandraZapuc**

### CODE REVIEW

File: rs/embedders/src/wasmtime_embedder/system_api_complexity.rs [hunk 1]
```diff
@@ -81,6 +81,11 @@ pub mod overhead {
     pub const STABLE64_READ: NumInstructions = NumInstructions::new(20);
     pub const STABLE64_SIZE: NumInstructions = NumInstructions::new(20);
     pub const STABLE64_WRITE: NumInstructions = NumInstructions::new(20);
+    pub const ENV_VAR_COUNT: NumInstructions = NumInstructions::new(20);
```
Comment (AlexandraZapuc) [inline rs/embedders/src/wasmtime_embedder/system_api_complexity.rs]: TODO different PR: https://dfinity.atlassian.net/browse/EXC-2063.

File: rs/embedders/src/wasmtime_embedder/system_api.rs [hunk 2]
```diff
@@ -1860,6 +1860,254 @@ impl SystemApi for SystemApiImpl {
...
+                        suggestion:
+                            "Provide a valid UTF-8 string for the environment variable name."
+                                .to_string(),
+                        doc_link: "".to_string(),
+                    }
+                })?;
+                match &self
+                    .sandbox_safe_system_state
+                    .environment_variables()
+                    .get(name)
```
Comment (AlexandraZapuc) [inline rs/embedders/src/wasmtime_embedder/system_api.rs:2065]: @mraszyk This is the scenario I was referring to—where we return an error if the name isn’t found.

Comment (mraszyk) [reply rs/embedders/src/wasmtime_embedder/system_api.rs:2065]: I think that's fine if there's an additional system API, e.g., `ic0.env_var_name_exists` to avoid this error in practice.

**Review by schneiderstefan — APPROVED** [review]
LGTM with the assumption that loading the variables from checkpoints will come before you can actually set these variables.

File: rs/embedders/src/wasmtime_embedder/linker.rs [hunk 3]
```diff
@@ -574,6 +574,98 @@ pub fn syscalls<
...
+        .unwrap();
+
+    linker
+        .func_wrap("ic0", "env_var_value_copy", {
+            move |mut caller: Caller<'_, StoreData>,
+                  dst: I,
+                  offset: I,
+                  size: I,
+                  name_src: I,
+                  name_size: I| {
```
Comment (mraszyk) [inline rs/embedders/src/wasmtime_embedder/linker.rs]: They should be put first.

File: rs/embedders/src/wasmtime_embedder/system_api.rs [hunk 4]
```diff
@@ -1860,6 +1860,254 @@ impl SystemApi for SystemApiImpl {
...
+            | ApiType::SystemTask { .. }
+            | ApiType::Update { .. }
+            | ApiType::Cleanup { .. }
+            | ApiType::NonReplicatedQuery { .. }
+            | ApiType::ReplicatedQuery { .. }
+            | ApiType::PreUpgrade { .. }
+            | ApiType::ReplyCallback { .. }
+            | ApiType::RejectCallback { .. }
+            | ApiType::InspectMessage { .. } => {
+                let name_bytes = valid_subslice(
```
Comment (mraszyk) [inline rs/embedders/src/wasmtime_embedder/system_api.rs:2045]: Shouldn't we also sanitize the length before performing a lookup with a potentially very long string?

File: rs/embedders/src/wasmtime_embedder/system_api.rs [hunk 5]
```diff
@@ -1860,6 +1860,254 @@ impl SystemApi for SystemApiImpl {
...
+                    "ic0.env_var_value_copy name",
+                    InternalAddress::new(name_src),
+                    InternalAddress::new(name_size),
+                    heap,
+                )?;
+
+                let name = std::str::from_utf8(name_bytes).map_err(|_| {
+                    HypervisorError::UserContractViolation {
+                        error: "ic0.env_var_value_copy: the name is not a valid UTF-8 string."
```
Comment (mraszyk) [inline rs/embedders/src/wasmtime_embedder/system_api.rs]:
```suggestion
                        error: "ic0.env_var_value_copy: Variable name is not a valid UTF-8 string."
```
for consistency

File: rs/embedders/src/wasmtime_embedder/system_api.rs [hunk 6]
```diff
@@ -1860,6 +1860,254 @@ impl SystemApi for SystemApiImpl {
...
+                    None => Err(EnvironmentVariableNotFound {
+                        name: name.to_string(),
+                    }),
+                }
+            }
+        };
+
+        trace_syscall!(
+            self,
+            EnvVarNameCopy,
```
Comment (mraszyk) [inline rs/embedders/src/wasmtime_embedder/system_api.rs]:
```suggestion
            EnvVarValueCopy,
```

File: rs/embedders/src/wasmtime_embedder/system_api_complexity.rs (hunk 1 above)
Comment (mraszyk) [reply rs/embedders/src/wasmtime_embedder/system_api_complexity.rs]: Maybe let's already start with 500 which is the value used for most system APIs.

File: rs/embedders/tests/system_api.rs [hunk 7]
```diff
@@ -780,6 +789,62 @@ fn api_availability_test(
...
+                api_type_enum,
+                context,
+            );
+        }
+        SystemApiCallId::EnvVarValueCopy => {
+            let mut heap = vec![0u8; 64];
+            let var_name = b"TEST_VAR_1";
+            copy_to_heap(&mut heap, var_name);
+            assert_api_availability(
+                |api| api.ic0_env_var_value_copy(0, 10, 0, 0, 0, &mut heap.clone()),
```
Comment (mraszyk) [inline rs/embedders/tests/system_api.rs]:
```suggestion
                |api| api.ic0_env_var_value_copy(0, var_name.size(), 0, 0, 0, &mut heap.clone()),
```

File: rs/embedders/tests/system_api.rs [hunk 8]
```diff
@@ -2191,3 +2256,182 @@ fn get_system_api_for_best_effort_response(
...
+
+    let mut heap = vec![0u8; 64];
+
+    // Test getting value size for existing variable
+    let var_name = b"TEST_VAR_1";
+    copy_to_heap(&mut heap, var_name);
+    assert_eq!(
+        api.ic0_env_var_value_size(0, var_name.len(), &heap)
+            .unwrap(),
+        11 // length of "Hello World"
```
Comment (mraszyk) [inline rs/embedders/tests/system_api.rs]: let's introduce a constant for "Hello World"

File: rs/embedders/tests/system_api.rs [hunk 9]
```diff
@@ -2191,3 +2256,182 @@ fn get_system_api_for_best_effort_response(
...
+    assert_eq!(&heap[0..11], b"Hello World");
+
+    // Test empty variable
+    let empty_var = b"EMPTY_VAR";
+    copy_to_heap(&mut heap, empty_var);
+    assert_eq!(
+        api.ic0_env_var_value_size(0, empty_var.len(), &heap)
+            .unwrap(),
+        0
+    );
```
Comment (mraszyk) [inline rs/embedders/tests/system_api.rs:2403]: Add test copying an empty variable.

File: rs/embedders/tests/system_api.rs [hunk 10]
```diff
@@ -2191,3 +2256,182 @@ fn get_system_api_for_best_effort_response(
...
+            name: "NON_EXISTENT".to_string()
+        })
+    );
+
+    // Test invalid UTF-8 in variable name
+    let invalid_utf8 = &[0xFF, 0xFF];
+    copy_to_heap(&mut heap, invalid_utf8);
+    assert!(matches!(
+        api.ic0_env_var_value_size(0, invalid_utf8.len(), &heap),
+        Err(HypervisorError::UserContractViolation { .. })
```
Comment (mraszyk) [inline rs/embedders/tests/system_api.rs]: let's also assert that the error is due to UTF-8

File: rs/embedders/tests/system_api.rs [hunk 11]
```diff
@@ -2191,3 +2256,182 @@ fn get_system_api_for_best_effort_response(
...
+    );
+    api.ic0_env_var_value_copy(0, path_var.len(), 0, 0, 23, &mut heap)
+        .unwrap();
+    assert_eq!(&heap[0..23], b"/usr/local/bin:/usr/bin");
+
+    // Test non-existent variable
+    let non_existent = b"NON_EXISTENT";
+    copy_to_heap(&mut heap, non_existent);
+    assert_eq!(
+        api.ic0_env_var_value_size(0, non_existent.len(), &heap),
```
Comment (mraszyk) [inline rs/embedders/tests/system_api.rs:2430]: let's also add the same test for env_var_value_copy

File: rs/interfaces/src/execution_environment.rs [hunk 12]
```diff
@@ -650,6 +660,41 @@ pub trait SystemApi {
...
+
+    /// Returns the size of the value for the environment variable with the given name.
+    fn ic0_env_var_value_size(
+        &self,
+        name_src: usize,
+        name_size: usize,
+        heap: &[u8],
+    ) -> HypervisorResult<usize>;
+
+    /// Copies the value of the environment variable with the given name into memory.
```
Comment (mraszyk) [inline rs/interfaces/src/execution_environment.rs:701]: Let's also point out the trap conditions (name is too long, name is not UTF-8, env var with given name does not exist).

File: rs/interfaces/src/execution_environment/errors.rs [hunk 13]
```diff
@@ -383,6 +390,16 @@ impl std::fmt::Display for HypervisorError {
                         limit.get(), bytes.get()
                 )
             }
+            Self::EnvironmentVariableIndexOutOfBounds { index, length } => {
+                write!(
+                    f,
+                    "Environment variable index {} is out of bounds. Length is {}.",
```
Comment (mraszyk) [inline rs/interfaces/src/execution_environment/errors.rs]:
```suggestion
                    "Environment variable index {} is out of bounds. The number of environment variables is {}.",
```

File: rs/embedders/src/wasmtime_embedder/system_api.rs (hunk 4 above)
Comment (AlexandraZapuc) [reply rs/embedders/src/wasmtime_embedder/system_api.rs:2045]: I have a task for adding validation for these conditions, both here and as part of the settings validation. I’ll go ahead and add it now with some initial limits. Later, when I implement the settings validation (which will use the same limits), I’ll revisit this.
I’d like to centralise these limits in one place to avoid duplication.

File: rs/embedders/src/wasmtime_embedder/linker.rs [hunk 14]
```diff
@@ -642,16 +642,16 @@ pub fn syscalls<
     linker
         .func_wrap("ic0", "env_var_value_copy", {
             move |mut caller: Caller<'_, StoreData>,
+                  name_src: I,
+                  name_size: I,
```
Comment (mraszyk) [inline rs/embedders/src/wasmtime_embedder/linker.rs]: The fact that CI was green before suggests we're lacking test coverage, e.g., in this module: rs/embedders/tests/wasmtime_embedder.rs.

**Review by mraszyk — COMMENTED** [review]
Since this PR does not ship the complete feature, could you please introduce a feature flag until the feature is completed?

File: rs/embedders/tests/system_api.rs [hunk 15]
```diff
@@ -2191,3 +2256,237 @@ fn get_system_api_for_best_effort_response(
...
+            index: 2,
+            length: 2
+        })
+    ));
+
+    // Test ic0_env_var_name_copy
+    let mut heap = vec![0u8; 16];
+
+    // Copy first variable name
+    api.ic0_env_var_name_copy(0, 0, 0, 10, &mut heap).unwrap();
```
Comment (mraszyk) [inline rs/embedders/tests/system_api.rs]:
```suggestion
    api.ic0_env_var_name_copy(0, 0, 0, var_name_1.len(), &mut heap).unwrap();
```
also below and for `var_name_2`

File: rs/embedders/tests/system_api.rs [hunk 16]
```diff
@@ -2191,3 +2256,237 @@ fn get_system_api_for_best_effort_response(
...
+        cycles_account_manager,
+    );
+    let mut heap = vec![0u8; 256];
+
+    // Test ic0_env_var_count.
+    assert_eq!(api.ic0_env_var_count().unwrap(), 3);
+
+    // Test copying value for null bytes as variable name.
+    assert_eq!(
+        api.ic0_env_var_value_copy(0, var_name_empty.len(), 0, 0, 0, &mut heap),
```
Comment (mraszyk) [inline rs/embedders/tests/system_api.rs]: I'd repeat the same test for var_value_size

File: rs/embedders/tests/system_api.rs [hunk 17]
```diff
@@ -2191,3 +2256,237 @@ fn get_system_api_for_best_effort_response(
...
+    api.ic0_env_var_value_copy(
+        0,
+        var_name_path.len(),
+        0,
+        0,
+        var_value_path.len(),
+        &mut heap,
+    )
+    .unwrap();
+    assert_eq!(&heap[0..23], var_value_path.as_bytes());
```
Comment (mraszyk) [inline rs/embedders/tests/system_api.rs]: To reduce code duplication, I'd refactor the three tests into a for loop over the three env variables with a uniform body handling the checks for the individual env vars.

File: rs/embedders/tests/system_api.rs [hunk 18]
```diff
@@ -2191,3 +2256,237 @@ fn get_system_api_for_best_effort_response(
...
+            .environment_variables(env_vars)
+            .build(),
+        cycles_account_manager,
+    );
+    let mut heap = vec![0u8; 256];
+
+    // Test ic0_env_var_count.
+    assert_eq!(api.ic0_env_var_count().unwrap(), 3);
+
+    // Test copying value for null bytes as variable name.
```
Comment (mraszyk) [inline rs/embedders/tests/system_api.rs]: This is also tested below requesting the name "NON_EXISTENT" so I'd drop one of those tests.

File: rs/embedders/tests/system_api.rs (hunk 17 above)
Comment (AlexandraZapuc) [reply rs/embedders/tests/system_api.rs]: [Ticket](https://dfinity.atlassian.net/browse/EXC-2070)

File: rs/embedders/src/wasmtime_embedder/linker.rs [hunk 19]
```diff
@@ -578,11 +578,17 @@ pub fn syscalls<
         .func_wrap("ic0", "env_var_count", {
             move |mut caller: Caller<'_, StoreData>| {
                 charge_for_cpu(&mut caller, overhead::ENV_VAR_COUNT)?;
-                with_system_api(&mut caller, |s| s.ic0_env_var_count()).and_then(|s| {
-                    I::try_from(s).map_err(|e| {
-                        anyhow::Error::msg(format!("ic0::env_var_count failed: {}", e))
+                if feature_flags.environment_variables == FlagStatus::Enabled {
```
Comment (mraszyk) [inline rs/embedders/src/wasmtime_embedder/linker.rs]: this "if" should be around `linker.func_wrap` so that we don't even allow the import of the system API for environment variables

File: rs/embedders/tests/system_api.rs [hunk 20]
```diff
@@ -2297,11 +2297,13 @@ fn test_env_var_name_operations() {
     let mut heap = vec![0u8; 16];
 
     // Copy first variable name
-    api.ic0_env_var_name_copy(0, 0, 0, 10, &mut heap).unwrap();
+    api.ic0_env_var_name_copy(0, 0, 0, var_name_1.len(), &mut heap)
+        .unwrap();
     assert_eq!(&heap[0..10], var_name_1.as_bytes());
```
Comment (mraszyk) [inline rs/embedders/tests/system_api.rs]:
```suggestion
    assert_eq!(&heap[0..var_name_1.len()], var_name_1.as_bytes());
```
also below, please

File: rs/embedders/tests/system_api.rs [hunk 21]
```diff
@@ -2315,13 +2317,19 @@ fn test_env_var_name_operations() {
...
+    // Test dst is not aligned
+    assert!(matches!(
+        api.ic0_env_var_name_copy(0, 10, 0, var_name_1.len(), &mut heap),
         Err(HypervisorError::ToolchainContractViolation { .. })
     ));
 
     // Test invalid size (destination buffer overflow)
     assert!(matches!(
-        api.ic0_env_var_name_copy(0, 10, 0, 10, &mut heap),
+        api.ic0_env_var_name_copy(0, 0, 0, 20, &mut heap),
```
Comment (mraszyk) [inline rs/embedders/tests/system_api.rs:2346]: Why this change?

Comment (mraszyk) [reply rs/embedders/tests/system_api.rs:2346]: i.e., does the test behavior changes for the values `0, 10, 0, 10` that were used before?

File: rs/embedders/tests/wasmtime_embedder.rs [hunk 22]
```diff
@@ -3418,3 +3418,125 @@ fn wasm64_saturate_fun_index() {
...
+        Ok(Some(WasmResult::Reply(b"Hello WorldTest Value".to_vec())))
+    );
+}
+
+// TODO(EXC-2071): Delete test when feature flag is removed.
+#[test]
+fn test_environment_variable_system_api_not_enabled() {
+    let wat = r#"
+    (module
+        (import "ic0" "env_var_count" (func $ic0_env_var_count (result i32)))
```
Comment (mraszyk) [inline rs/embedders/tests/wasmtime_embedder.rs:3506]: This should fail to install if the feature flag is disabled.

File: rs/embedders/tests/wasmtime_embedder.rs [hunk 23]
```diff
@@ -3418,3 +3418,125 @@ fn wasm64_saturate_fun_index() {
...
+        (call $ic0_env_var_count)
+        drop
+
+        ;; Get and print first variable
+        ;; Get name size: 10.
+        (call $env_var_name_size (i32.const 0))
+        drop
+
+        ;; Copy first name to memory
+        (call $env_var_name_copy (i32.const 0) (i32.const 0) (i32.const 0) (i32.const 10))
```
Comment (mraszyk) [inline rs/embedders/tests/wasmtime_embedder.rs]:
```suggestion
        (call $env_var_name_copy (i32.const 0) (i32.const 0) (i32.const 0) (call $env_var_name_size (i32.const 0)))
```

File: rs/embedders/tests/wasmtime_embedder.rs [hunk 24]
```diff
@@ -3418,3 +3418,125 @@ fn wasm64_saturate_fun_index() {
...
+      (import "ic0" "env_var_count" (func $ic0_env_var_count (result i32)))
+      (import "ic0" "env_var_name_size" (func $env_var_name_size (param i32) (result i32)))
+      (import "ic0" "env_var_name_copy" (func $env_var_name_copy (param i32 i32 i32 i32)))
+      (import "ic0" "env_var_value_size" (func $env_var_value_size (param i32 i32) (result i32)))
+      (import "ic0" "env_var_value_copy" (func $env_var_value_copy (param i32 i32 i32 i32 i32)))
+    
+            
+      (func (export "canister_update go")
+        (call $ic0_env_var_count)
+        drop
```
Comment (mraszyk) [inline rs/embedders/tests/wasmtime_embedder.rs:3437]: We could append this as the first byte of the reply instead of dropping.

File: rs/embedders/tests/wasmtime_embedder.rs (hunk 23 above)
Comment (mraszyk) [reply rs/embedders/tests/wasmtime_embedder.rs]: then we could drop the unused call above; same below for `env_var_value_size`

File: rs/embedders/tests/system_api.rs (hunk 20 above)
Comment (mraszyk) [reply rs/embedders/tests/system_api.rs]: The goal would be to have no integer literals in the test: they are hard to make sense of later on if the test needs to be debugged/refactored.

File: rs/embedders/tests/system_api.rs (hunk 21 above)
Comment (AlexandraZapuc) [reply rs/embedders/tests/system_api.rs:2346]: Is there a specific reason you’re against checking these different inputs in the test? At this point, the level of nitpicking is becoming unproductive and isn’t adding value to the review.

I’ve included a test for an invalid size and one for the dst input, which triggers some of the slice validation logic.

File: rs/embedders/tests/wasmtime_embedder.rs (hunk 22 above)
Comment (AlexandraZapuc) [reply rs/embedders/tests/wasmtime_embedder.rs:3506]: Fair point.

File: rs/embedders/tests/wasmtime_embedder.rs (hunk 23 above)
Comment (AlexandraZapuc) [reply rs/embedders/tests/wasmtime_embedder.rs]: Sure, this is actually how I had it initially, but I quickly realized that this approach makes the code less readable.

To be straightforward: I’m happy to address the nitpick as you prefer, but I’ll apply the change only to the copy calls — not to the other instances.

File: rs/embedders/tests/wasmtime_embedder.rs (hunk 24 above)
Comment (AlexandraZapuc) [reply rs/embedders/tests/wasmtime_embedder.rs:3437]: Once again, I believe this is a bit of a nitpick.  The function is already covered in other test suites, so I don’t see this as a particularly critical addition.

File: rs/embedders/tests/system_api.rs (hunk 21 above)
Comment (mraszyk) [reply rs/embedders/tests/system_api.rs:2346]: > Is there a specific reason you’re against checking these different inputs in the test?

I just want to understand what changed and if we need to improve test coverage. The previous values seemed to be fine so I wonder what triggered the change.

File: rs/embedders/tests/wasmtime_embedder.rs (hunk 24 above)
Comment (mraszyk) [reply rs/embedders/tests/wasmtime_embedder.rs:3437]: I agree that this is less critical since this API is very simple. But having it covered seems like a good idea to me. Anyway, I'm not gonna block this PR on that so feel free to resolve if you don't want to cover that API in the integration test.

File: rs/embedders/tests/wasmtime_embedder.rs (hunk 23 above)
Comment (mraszyk) [reply rs/embedders/tests/wasmtime_embedder.rs]: > I’ll apply the change only to the copy calls — not to the other instances

I think this is what I meant: use the size returned by the size calls in the copy calls (and then drop the separate size calls).

Comment (mraszyk) [reply rs/embedders/tests/wasmtime_embedder.rs]: > I quickly realized that this approach makes the code less readable.

I'd recommend to split the code into multiple lines (one term per line) with comments on the individual terms.

**Comment by AlexandraZapuc** [comment]
@mraszyk I added the code changes to address your last comments. Is there anything else that is blocking the approval of this PR?

**Review by mraszyk — APPROVED** [review]

File: rs/embedders/tests/system_api.rs (hunk 21 above)
Comment (AlexandraZapuc) [reply rs/embedders/tests/system_api.rs:2346]: As I mentioned that one is triggering the slice validation check because of the dst value, another one because of the size value.
